### PR TITLE
Kaleh and Nym AMLAs

### PR DESCRIPTION
**Pull request by Hejnewar**
So I finally got some decent concept for doing this. General pourpose of this is making them more interesting, customisable and original as well as making playthrus more of a unique experience. 

Legend:
Arrows - amla required
Orance colour - exclusive almas, you can only get one.

New Kaleh ALMA tree: 
New Kaleh (1)

New Nym ALMA tree: 
New Nym

Hmm... even now, I wanted to write something about goals or something grand, but whatever, I still see thing to improve, like adding at least one AMLA to Kaleh melee tree and bola disengage for Nym. Discuss!

### CODE REVIEW

**Comment by Pentarctagon** [comment]
It looks like Overwhelming Power and Stronger Grip do the same thing?

**Comment by Hejnewar** [comment]
For now yes, because the PR has no abilities for now because I really didnt want it to be about code or balalance just concepts of the AMLAs.

If you look closely, it has W_S_PIERCE 2, and W_S_COOLDOWN 2, standard Piercing Shot AMLA has W_S_PIERCE, and W_S_COOLDOWN 1, generally unlike stronger grip which is supposed to be simple upgrade to the attack Overwhelming Power is supposed to have way more power and bigger drawback. For now its just cooldown, but Im still thinking if I should or shouldnt give him some additional penalties after using it. 

Also the numbers are way off cuz there is no balance there yet. 

Also also, after tesing most of that, Taunt AMLA will be the most problematic I feel like because there is no attack and follow micro ai there are prarts of it I might get it to work with two mais but I fear that it might be sometimes unreliable, we will see, Fear seems simpler to do right but doesnt fit Kaleh IMO.

**Comment by Pentarctagon** [comment]
What do the green vs black outlines mean?

**Comment by Hejnewar** [comment]
Nothing really, basically what I have done / what I don't have done. At this point I have done everything except for one Amla (cuz I didn't expect they would be that fast to do).

**Comment by nemaara** [comment]
Kaleh's looks pretty good to me if you can get taunt to work properly. Numbers can be tweaked later. Only real comment is that I don't really like his swift tree, since he's supposed to be a bit slower and more about support or power (character-wise) than Nym. So instead of skirmisher, maybe he could have distract, and instead of fast, he could get a "sand cloud" ability that reduces adjacent enemy moves on start of turn (or even slows). Also swift being prerequisite for tough doesn't really thematically make sense.

For Nym I'm slightly worried that sword dance would be underpowered for how many prerequisites it has (but again values can be adjusted later). Arrow dancing is a weird name, maybe it can be called something simple like "ranger" instead? No other comments on rest of tree, it looks pretty good.

**Comment by CelticMinstrel** [comment]
The main thing that seems weird to me is "Bolaship", which based on its prerequisites seems like it should just be called "Opportunist". Even if that doesn't work, I don't think "Bolaship" is a good name at all.

**Comment by Hejnewar** [comment]
First response, Kaleh is pretty skinny on his art so I dont mind, I will remove fast, I actually need some space in the tree. I will look into tough and swift, I might even swap stength training and tough in places we will see I will think about it. 

Also keep in mind that I will cut down on basic stats of theirs. I actually dont think I included that here for some reason, the reason for that is actually making the first 4-5 amlas simpler and more of a gateway because if I started from tier 2 amlas I had a feeling that some players could be overwhelmed by them (I might also do nightmare / normal versions but I dont think it will be necessary). 

The ability names are like whatever to me at this point even icons are more important, but I dont want to have 2 amlas in the same campaign that do different things but have the same name and Opportunist is way better for the ability of Kaleh.

**Comment by CelticMinstrel** [comment]
> The ability names are like whatever to me at this point even icons are more important, but I dont want to have 2 amlas in the same campaign that do different things but have the same name and Opportunist is way better for the ability of Kaleh.

It's also kinda weird that they do different things when _all_ of their prerequisites are completely identical.

**Comment by Hejnewar** [comment]
> It's also kinda weird that they do different things when all of their prerequisites are completely identical. 

Nah thats not weird at all people can train in the same things but develop distinct styles anyway.

**Comment by Hejnewar** [comment]
What a race against time, finished barely on time but Im happy. No time to even write a lot. 

Improved Kaleh:
New Kaleh (2)
Improved Nym:
New Nym (1)

Nym will get like one more amla.

File: data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg [hunk 1]
```diff
@@ -1319,14 +1319,513 @@ Marksman attacks are only affected if the chance to hit is greater than 60%."
...
+
+        {CLEAR_VARIABLE opportunist_hit_units,opportunist_hit_chance,opportunist_hit_roll}
+    [/event]
+#enddef
+
+#define WEAPON_SPECIAL_TAUNT
+    [dummy]
+        id = taunt
+        name = _"taunt"
+        description= _ "It taunts the enemy making it try to attack Kaleh no matter what."
```
Comment (stevecotton) [inline data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg:1401]: The pronoun "it" appears twice here, applying to different things.

File: data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg [hunk 2]
```diff
@@ -1319,14 +1319,513 @@ Marksman attacks are only affected if the chance to hit is greater than 60%."
...
+#enddef
+
+#define STR_COOLDOWN_NAME_2
+_ "cooldown 1"#enddef
+
+#define STR_COOLDOWN_NAME_3
+_ "cooldown 2"#enddef
+
+#define STR_COOLDOWN_RESTRICT_2
+_ "After using this attack, you cant use it during your next turn"#enddef
```
Comment (stevecotton) [inline data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg:1463]: can't

File: data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg [hunk 3]
```diff
@@ -1319,14 +1319,513 @@ Marksman attacks are only affected if the chance to hit is greater than 60%."
...
+
+#define STR_COOLDOWN_RESTRICT_3
+_ "After using this attack, you cant use it during your next two turn"#enddef
```
Comment (stevecotton) [inline data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg:1466]: "can't" and "turns"

File: data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg [hunk 4]
```diff
@@ -1319,14 +1319,513 @@ Marksman attacks are only affected if the chance to hit is greater than 60%."
...
+
+#define WEAPON_SPECIAL_COOLDOWN X
+    [disable]
+        id = cooldown_{X}
+        name = {STR_COOLDOWN_NAME_{X}}
+        description = {STR_COOLDOWN_RESTRICT_{X}}
+        [filter_self]
+            formula = "self.wml_vars.cooldown > 0"
```
Comment (stevecotton) [inline data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg:1474]: This copies the design of the song abilities - it's a per-weapon cooldown, but put in a per-unit variable. If two weapons have cooldowns, they'll conflict with each other.

Does this one have a lore reason why it can be used on defense even when it's on cooldown?

File: data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg [hunk 5]
```diff
@@ -1319,14 +1319,513 @@ Marksman attacks are only affected if the chance to hit is greater than 60%."
...
+                [/modify_unit]
+            [/do]
+        [/foreach]
+    [/event]
+#enddef
+
+#define WEAPON_SPECIAL_PIERCE
+    [dummy]
+        id = pierce
+        name = _"pierce"
```
Comment (stevecotton) [inline data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg:1535]: Uses the name of one of the 6 basic unit types for a different purpose, which is confusing.

For the same reason, I'd expect `WEAPON_SPECIAL_PIERCE` to change the damage type, not do something unrelated.

File: data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg [hunk 6]
```diff
@@ -1319,14 +1319,513 @@ Marksman attacks are only affected if the chance to hit is greater than 60%."
...
+        description= _ "This weapon, when used offensively, also deals damage to a unit behind the target."
```
Comment (stevecotton) [inline data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg:1536]: "can also deal damage", as it's random

File: data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg [hunk 7]
```diff
@@ -1319,14 +1319,513 @@ Marksman attacks are only affected if the chance to hit is greater than 60%."
...
+
+        {CLEAR_VARIABLE puncture_hit_units,puncture_hit_chance,puncture_hit_roll}
+    [/event]
+#enddef
+
+#define WEAPON_SPECIAL_RAIN_OF_ARROWS
+    [dummy]
+        id = rain_of_arrows
+        name = _"rain of arrows"
+        description= _ "This weapon, when used offensively, also deals damage to a unit behind the target and all units adjacent to the unit behin the target."
```
Comment (stevecotton) [inline data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg:1675]: "behin"

File: data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg [hunk 8]
```diff
@@ -1319,14 +1319,513 @@ Marksman attacks are only affected if the chance to hit is greater than 60%."
...
+
+        {CLEAR_VARIABLE rain_of_arrows_hit_units,rain_of_arrows_hit_chance,rain_of_arrows_hit_roll}
+    [/event]
+#enddef
+
+#define WEAPON_SPECIAL_SWORD_DANCE
+    [dummy]
+        id = sword_dance
+        name = _"sword dance"
+        description= _ "Every time Nym hits an enemy she gains +10% accuracy and 10% parry."
```
Comment (stevecotton) [inline data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg:1783]: Until when?

File: data/campaigns/Under_the_Burning_Suns/units/quenoth/Youth.cfg [hunk 9]
```diff
@@ -30,291 +31,638 @@ _"Nym is Kaleh’s childhood friend, a young, rebellious lady with quick wits an
...
             [abilities]
-                {ABILITY_SUPPORT}
+                {ABILITY_OPPORTUNIST} 
             [/abilities]
         [/effect]
+    [/advancement]
+
+    [advancement]
+        id = insipring
+        description = _"Insipring"
```
Comment (stevecotton) [inline data/campaigns/Under_the_Burning_Suns/units/quenoth/Youth.cfg:360]: Spell-check

File: data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg (hunk 1 above)
Comment (Zireael07) [reply data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg:1401]: Not to mention the initial "it" looks superfluous to me...

**Comment by stevecotton** [comment]
Seems a very low ratio of comments to the amount of code added. My experience is, taking the time to write comments means you spend less time debugging afterwards.

File: data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg (hunk 1 above)
Comment (CelticMinstrel) [reply data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg:1401]: I agree, you can just drop the initial "it". That said, the second "it" should maybe be "them" since it refers to a person.

File: data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg (hunk 5 above)
Comment (CelticMinstrel) [reply data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg:1535]: Maybe "penetrate"?

File: data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg (hunk 4 above)
Comment (gfgtdf) [reply data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg:1474]: So you have a suggestion on what to do about the weapon cooldown? Do you want to add [variables] in attacks? Or maybe use the name of the attack as a variable key?

Comment (stevecotton) [reply data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg:1474]: Documenting the limitation in comments would help a lot, even if it doesn't change the implementation.

File: data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg [hunk 10]
```diff
@@ -1319,14 +1319,513 @@ Marksman attacks are only affected if the chance to hit is greater than 60%."
...
+        first_time_only = no
+
+        [store_unit]
+            [filter]
+                [has_attack]
+                    special_id = cooldown_{X}
+                [/has_attack]
+            [/filter]
+
+            variable = cooling_units
```
Comment (stevecotton) [inline data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg:1512]: Why does this need to use `[store_unit]` and a loop instead of just putting this filter into `[modify_unit]`?

I'd understand if it was checking for units with `cooldown > 0` and only applying `[modify_unit]` to ones that are cooling down, but it isn't.

File: data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg (hunk 4 above)
Comment (Hejnewar) [reply data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg:1474]: > Does this one have a lore reason why it can be used on defense even when it's on cooldown?

It cant, the wole attack using it cant be used on defense.

File: data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg (hunk 5 above)
Comment (Hejnewar) [reply data/campaigns/Under_the_Burning_Suns/utils/abilities.cfg:1535]: "Kaleh penetrated 2 enemies with his (long) arrow" 

IMO pierce is fine

**Comment by Hejnewar** [comment]
Tbh if you look at it, Opportunist, Rain of Arrow, Pierce and Puncture are pretty much the same ability with different filter, Taunt and Sword Dance are very similar and work on the same basics, Cooldown is kinda not copied. The amount of actuall code is small, I just copy it a lot with small changes. 

Reuse of assets ftw.

**Comment by Hejnewar** [comment]
Two problems, 1) Ranger and Sword Dance amlas dont work as uses attacks special is not replaced even when I try to remove it fisrt. 
2) Need new name for 0 movement attacks I think cuz there is ability Disengage and it does something lil bit differernt.

sleep

**Comment by Hejnewar** [comment]
Sorry for not responding a lot but Im basically doing this whenever I have time before sleep so like every other day. But I resolve convos of whatever I fixed. 

New Nym again: 
New Nym (2)

Im starting the playtests right now not for bugs but for the feeling. So far Im happy.

About taunt:
Honestly I never wanted it to appear in MP, I tried to apply mocro ai to all possible sides at all times but if a side doesnt exist then it cant be applied, that triggers a message that ultimately I didnt want players to see. 

In custom scenarios it will work pretty much as is. In custom factions I dont really want to see it, cuz sometimes you fight other players with them and then its useless.

**Comment by CelticMinstrel** [comment]
> About taunt: Honestly I never wanted it to appear in MP, I tried to apply mocro ai to all possible sides at all times but if a side doesnt exist then it cant be applied, that triggers a message that ultimately I didnt want players to see.
> 
> In custom scenarios it will work pretty much as is. In custom factions I dont really want to see it, cuz sometimes you fight other players with them and then its useless.

It's not technically relevant to this PR, yeah, as UtBS is a single-player scenario and only Nym and Kaleh have access to it. But conceptually I think it would be nice to conceive of a taunt that works in MP.

**Comment by Hejnewar** [comment]
Two concerning things, I noticed that enemy units sometimes do nothing if player or enemy is oustide of their range if their side has no leader. 

I noticed that healing doesnt always trigger, especially when using healer train (im big fan of it for s2). Its often just one healer healing.

File: data/campaigns/Under_the_Burning_Suns/utils/macros.cfg [hunk 11]
```diff
@@ -136,7 +136,7 @@ if(red <= green and green <= blue + 3, ({NEWCHAN} / 255.0)^2 * 255.0, {ORIGCHAN}
 #enddef
 
 #define UTBS_XP_MODIFIER
-    experience_modifier={ON_DIFFICULTY 100 150 200}
+    experience_modifier={ON_DIFFICULTY 100 125 150}
```
Comment (nemaara) [inline data/campaigns/Under_the_Burning_Suns/utils/macros.cfg:139]: This balance change should be in a separate commit imo, along with the recall cost change below.

**Comment by nemaara** [comment]
No comment on tree, it looks fine to me. I can help with the naming later.

File: data/campaigns/Under_the_Burning_Suns/utils/macros.cfg (hunk 11 above)
Comment (Hejnewar) [reply data/campaigns/Under_the_Burning_Suns/utils/macros.cfg:139]: Probably.

**Comment by Hejnewar** [comment]
So is the code fine?

**Comment by Hejnewar** [comment]
So beasically this is failing because of not updated shema huh.

File: data/campaigns/Under_the_Burning_Suns/units/quenoth/Youth.cfg [hunk 12]
```diff
@@ -30,291 +31,707 @@ _"Nym is Kaleh’s childhood friend, a young, rebellious lady with quick wits an
...
+            increase_total = 4
+            heal_full = yes
         [/effect]
     [/advancement]
     [advancement]
-        id=leader_1
-        description= _ "Leader"
-        image=attacks/leader.png
+        id = run_by_taunt
+        description = _"Run By (Taunting strike)"
```
Comment (cooljeanius) [inline data/campaigns/Under_the_Burning_Suns/units/quenoth/Youth.cfg:145]: I've received word from a translator having trouble understanding this description; please add a `# po:` comment to explain how it works a bit better
